README.md
# True Color Tiles Microservice

[![Build Status](https://travis-ci.com/gfw-api/true-color-tiles.svg?branch=dev)](https://travis-ci.com/gfw-api/true-color-tiles)
[![Test Coverage](https://api.codeclimate.com/v1/badges/90a8a96a1b064267f42d/test_coverage)](https://codeclimate.com/github/gfw-api/true-color-tiles/test_coverage)

This repository moves the dynamic pixel showing/hiding/recoloring magic from the GFW flagship site to the server-side, making true color loss and GLAD tiles more easily available.

To demo this service, try adding the following to the map on http://geojson.io (Meta -> Add Map Layer):
http://production-api.globalforestwatch.org/v1/true-color-tiles/loss/{z}/{x}/{y}?thresh=75&startYear=2005&endYear=2008

The above will add loss tiles to the geojson.io map, filtering to show only data from 2005 - 2008 where thresh >=75.

This is the same mechanism that the flagship performs when users change the timeline and the TCD threshold of interest.

For an example of adding zoomable loss/glad tiles to ArcMap, see [Add-Data-to-ArcMap.md](Add-Data-to-ArcMap.md)

## Endpoints

### Annual Loss Layer
http://production-api.globalforestwatch.org/v1/true-color-tiles/loss/{z}/{x}/{y}

Query Parameters

| Parameter | Type | Description | 
|-----------|------|-------------|
|thresh| Int | Tree cover Density Threshold used to mask Tree cover loss (10/15/20/25/30/50/75 - default 30) |
|startYear | Int | Will filter out all loss prior to given year (default 2000) |
|endYear | Int | Will filter out all loss after given year (default 2018) |


### GLAD alerts
http://production-api.globalforestwatch.org/v1/true-color-tiles/glad/{z}/{x}/{y}

| Parameter | Type | Description | 
|-----------|------|-------------|
|startDate | Date | Will filter out all alerts prior to given date. Use format YYYY-MM-DD (default 2015-01-01) |
|endDate | Date | Will filter out all alerts after given date. Use format YYYY-MM-DD (default today) |
|showUnconfirmed | Bool | Show unconfirmed alerts (default False) |

### Running locally on OS X

**First, make sure that you have the [API gateway running
locally](https://github.com/control-tower/control-tower).**

We're using Docker which, luckily for you, means that getting the
application running locally should be fairly painless. First, make sure
that you have [Docker Compose](https://docs.docker.com/compose/install/)
installed on your machine.

```
git clone https://github.com/gfw-api/true-color-tiles
cd true-color-tiles
./true-color-tiles.sh develop
```


CHANGELOG.md
## 31/05/2021

- Update `rw-api-microservice-node` to add CORS support.

## 29/01/2021

- Replace CT integration library

# v1.0.0

## 08/04/2020

- Update k8s configuration with node affinity.


app/src/app.js
const Koa = require('koa');
const logger = require('logger');
const koaLogger = require('koa-logger');
const config = require('config');
const loader = require('loader');
const convert = require('koa-convert');
const { RWAPIMicroservice } = require('rw-api-microservice-node');
const ErrorSerializer = require('serializers/error.serializer');
const koaSimpleHealthCheck = require('koa-simple-healthcheck');

const koaBody = require('koa-body')({
    multipart: true,
    jsonLimit: '50mb',
    formLimit: '50mb',
    textLimit: '50mb'
});

const app = new Koa();

app.use(convert(koaBody));

app.use(koaSimpleHealthCheck());

app.use(async (ctx, next) => {
    try {
        await next();
    } catch (inErr) {
        let error = inErr;
        try {
            error = JSON.parse(inErr);
        } catch (e) {
            logger.debug('Could not parse error message - is it JSON?: ', inErr);
            error = inErr;
        }
        ctx.status = error.status || ctx.status || 500;
        if (ctx.status >= 500) {
            logger.error(error);
        } else {
            logger.info(error);
        }

        ctx.body = ErrorSerializer.serializeError(ctx.status, error.message);
        if (process.env.NODE_ENV === 'prod' && ctx.status === 500) {
            ctx.body = 'Unexpected error';
        }
        ctx.response.type = 'application/vnd.api+json';
    }
});

app.use(koaLogger());

app.use(RWAPIMicroservice.bootstrap({
    name: config.get('service.name'),
    info: require('../microservice/register.json'),
    swagger: require('../microservice/public-swagger.json'),
    logger,
    baseURL: process.env.CT_URL,
    url: process.env.LOCAL_URL,
    token: process.env.CT_TOKEN,
    fastlyEnabled: process.env.FASTLY_ENABLED,
    fastlyServiceId: process.env.FASTLY_SERVICEID,
    fastlyAPIKey: process.env.FASTLY_APIKEY
}));


loader.loadRoutes(app);

const server = app.listen(process.env.PORT, () => {
    if (process.env.CT_REGISTER_MODE === 'auto') {
        RWAPIMicroservice.register().then(() => {
            logger.info('CT registration process started');
        }, (error) => {
            logger.error(error);
            process.exit(1);
        });
    }
});

logger.info('Server started in ', process.env.PORT);

module.exports = server;


app/src/services/image.service.js
/* eslint-disable no-restricted-properties,no-param-reassign,no-mixed-operators,no-use-before-define,no-undef,no-bitwise */
const logger = require('logger');
const Canvas = require('canvas');
const d3 = require('d3');
const rp = require('request-promise');

const dataMaxZoom = 12;
const assert = require('assert');

class ImageService {

    static pad(num) {
        const s = `00${num}`;
        return s.substr(s.length - 3);
    }

    static decodeGLAD(data, ctx) {

        const q = ctx.query;

        const startDate = (q.startDate === undefined) ? 0 : ImageService.dateToInt(q.startDate, 0);
        const endDate = (q.endDate === undefined) ? 9999 : ImageService.dateToInt(q.endDate, 9999);
        const showUnconfirmed = (q.showUnconfirmed === undefined) ? false : ImageService.strToBool(q.showUnconfirmed);

        for (let i = 0; i < data.length; i += 4) {

            const band3Str = ImageService.pad(data[i + 2].toString());
            const confidence = parseInt(band3Str[0], 10) - 1;
            const totalDays = data[i] * 255 + data[i + 1];

            if ((totalDays > 0 && totalDays >= startDate && totalDays <= endDate) && (confidence === 1 || showUnconfirmed)) {


                const intensityRaw = parseInt(band3Str.slice(1, 3), 10);
                let intensity = intensityRaw * 50;

                if (intensity > 255) {
                    intensity = 255;
                }

                data[i] = 220;
                data[i + 1] = 102;
                data[i + 2] = 153;
                data[i + 3] = intensity;

            } else {
                data[i + 3] = 0;
            }
        }

        return data;
    }


    static strToBool(boolStr) {
        if (boolStr.toLowerCase() === 'true' || boolStr.toLowerCase() === 'yes' || boolStr === '1') return true;
        return false;

    }


    static dateToInt(yearStr, defaultvalue) {

        function isLeapYear(year) {
            return !(year % 4);
        }

        function dayOfYear() {
            const dayCount = [0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334];

            let dayOfYear = dayCount[month] + day;
            if (month > 1 && isLeapYear(year)) dayOfYear++;
            return dayOfYear;
        }

        function baseDays() {
            let leapYears = 0;
            for (let i = baseYear - 1; i < year; i++) {
                leapYears += isLeapYear(i) ? 1 : 0;
            }

            return (year - baseYear) * 365 + leapYears;
        }

        const baseYear = 2015;
        let result = defaultvalue;


        try {

            logger.debug(`yearStr: ${yearStr}`);

            const year = parseInt(yearStr.substring(0, 4), 10);
            const month = parseInt(yearStr.substring(5, 7), 10);
            const day = parseInt(yearStr.substring(8, 10), 10);

            assert(year >= baseYear);
            assert(month >= 1 && month <= 12);
            assert(day >= 1 && day <= 31);

            result = baseDays() + dayOfYear();

        } catch (e) {
            logger.debug(e);
        }

        return result;

    }

    static yearToInt(yearStr) {
        return parseInt(yearStr.slice(-2), 10);
    }

    static decodeLoss(data, ctx) {

        const { z } = ctx.params;
        const q = ctx.query;

        const startYear = (q.startYear === undefined) ? 0 : ImageService.yearToInt(q.startYear);
        const endYear = (q.endYear === undefined) ? 100 : ImageService.yearToInt(q.endYear);

        const exp = z < 11 ? 0.3 + ((z - 3) / 20) : 1;

        const myscale = d3.scalePow()
            .exponent(exp)
            .domain([0, 256])
            .range([0, 256]);

        for (let i = 0; i < data.length; i += 4) {

            if (endYear >= data[i + 2] && data[i + 2] >= startYear) {

                const intensity = data[i];

                data[i] = 220;
                data[i + 1] = (72 - z) + 102 - (3 * myscale(intensity) / z);
                data[i + 2] = (33 - z) + 153 - ((intensity) / z);
                data[i + 3] = z < 13 ? myscale(intensity) : intensity;

            } else {
                data[i + 3] = 0;
            }
        }

        return data;
    }

    static getUrl(urlTemplate, coords) {
        return urlTemplate.replace('%z', coords[2]).replace('%x', coords[1]).replace('%y', coords[0]);
    }

    static getZoomSteps(z) {
        return z - dataMaxZoom;
    }

    static getTileCoords(x, y, z) {
        if (z > dataMaxZoom) {
            x = Math.floor(x / (Math.pow(2, z - dataMaxZoom)));
            y = Math.floor(y / (Math.pow(2, z - dataMaxZoom)));
            z = dataMaxZoom;
        } else {
            y = (y > Math.pow(2, z) ? y % Math.pow(2, z) : y);
            if (x >= Math.pow(2, z)) {
                x %= Math.pow(2, z);
            } else if (x < 0) {
                x = Math.pow(2, z) - Math.abs(x);
            }
        }
        return [x, y, z];
    }

    static async getImage(reqCtx) {
        logger.info('Getting url');

        const { params } = reqCtx;
        const url = this.getUrl(params.urlTemplate, this.getTileCoords(params.x, params.y, params.z));

        const team = await
        rp({ url, encoding: null }, (err) => {
            if (err) throw err;

        }).then((body) => {

            const img = new Canvas.Image();

            // eslint-disable-next-line no-buffer-constructor
            img.src = new Buffer(body, 'binary');
            const canvas = new Canvas(img.width, img.height);

            const ctx = canvas.getContext('2d');

            const zsteps = ImageService.getZoomSteps(params.z) | 0; // force 32bit int type
            // ctx.clearRect(0, 0, 256, 256);                    // this will allow us to sum up the dots when the timeline is running
            if (zsteps < 0) {
                ctx.drawImage(img, 0, 0);
            } else { // over the maxzoom, we'll need to scale up each tile
                ctx.imageSmoothingEnabled = false; // disable pic enhancement
                ctx.mozImageSmoothingEnabled = false;

                // tile scaling
                const srcX = (256 / Math.pow(2, zsteps) * (params.x % Math.pow(2, zsteps))) | 0;
                const srcY = (256 / Math.pow(2, zsteps) * (params.y % Math.pow(2, zsteps))) | 0;
                const srcW = (256 / Math.pow(2, zsteps)) | 0;
                const srcH = (256 / Math.pow(2, zsteps)) | 0;
                ctx.drawImage(img, srcX, srcY, srcW, srcH, 0, 0, 256, 256);
            }
            const I = ctx.getImageData(0, 0, canvas.width, canvas.height);

            if (params.layer === 'glad') {
                ImageService.decodeGLAD(I.data, reqCtx);
            } else {
                ImageService.decodeLoss(I.data, reqCtx);
            }

            ctx.putImageData(I, 0, 0);

            return canvas.toBuffer();
        });
        return team;
    }

}

module.exports = ImageService;


app/src/routes/api/v1/service.router.js
const Router = require('koa-router');
const ImageService = require('services/image.service');
const logger = require('logger');


const router = new Router({
    prefix: '/true-color-tiles',
});

class Service {

    static async drawTile(ctx) {
        const url = 'http://storage.googleapis.com/wri-public/Hansen18/tiles/hansen_world/v1/tc%thresh/%z/%y/%x.png';

        switch (ctx.params.layer) {

            case 'glad':
                ctx.params.urlTemplate = 'http://tiles.globalforestwatch.org/glad_prod/tiles/%z/%y/%x.png';
                break;

            case 'loss':
                ctx.params.urlTemplate = url.replace('%thresh', Service.validateThresh(ctx));
                break;

            default:
                ctx.throw(400, 'Wrong layer parameter supplied, should be loss or glad');

        }

        let image;

        try {
            image = await ImageService.getImage(ctx);
            ctx.body = image;
        } catch (e) {
            logger.error(e);
            ctx.throw(500, 'Error while retrieving image');
        }

    }

    static validateThresh(ctx) {
        const thresh = (ctx.query.thresh === undefined) ? '30' : ctx.query.thresh;

        const threshVals = [10, 15, 20, 25, 30, 50, 75];
        const validThresh = threshVals.includes(parseInt(thresh, 10));

        if (!validThresh) {
            ctx.throw(`Thresh supplied not in ${threshVals}`);
        }

        return thresh;
    }

}

router.get('/:layer/:z/:x/:y', Service.drawTile);

module.exports = router;
